### ISSUE TITLE
OT 3.3.1 With Saxon-HE 9.8.0.15 Generates Invalid Catalog File

### ISSUE THREAD

**Issue by ToshihikoMakita** [dita-ot/dita-ot#3284]
DITA-OT 3.3.1 bundles Saxon-HE 9.8.0.14. When I change it to Saxon-HE 9.8.0.15, the `bin/dita --install` command generates invalid `catalog-dita.xml` with `xmlns=""`. As a result, all of the build attempts fails when XSLT stylesheet uses `plugin:` prefix.

[The result of dita-ot-3.3.1/plugins/org.dita.base/catalog-dita.xml with Saxon-HE 9.8.0.14]

    <?xml version="1.0" encoding="UTF-8"?><catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog" xmlns:dita="http://dita-ot.sourceforge.net">

      <rewriteURI uriStartString="platform:" rewritePrefix="../../"/>

      <rewriteURI uriStartString="plugin:org.dita.htmlhelp:" rewritePrefix="../org.dita.htmlhelp/"/><rewriteURI uriStartString="plugin:org.dita.base:" rewritePrefix="../org.dita.base/"/><rewriteURI uriStartString="plugin:org.dita.pdf2.xep:" rewritePrefix="../org.dita.pdf2.xep/"/><rewriteURI uriStartString="plugin:com.sophos.tocjs:" rewritePrefix="../com.sophos.tocjs/"/><rewriteURI uriStartString="plugin:org.dita.pdf2.fop:" rewritePrefix="../org.dita.pdf2.fop/"/><rewriteURI uriStartString="plugin:org.dita.eclipsehelp:" rewritePrefix="../org.dita.eclipsehelp/"/><rewriteURI uriStartString="plugin:org.dita.specialization.dita11:" rewritePrefix="../org.dita.specialization.dita11/"/><rewriteURI uriStartString="plugin:org.dita.normalize:" rewritePrefix="../org.dita.normalize/"/><rewriteURI uriStartString="plugin:org.oasis-open.xdita.v0_2_2:" rewritePrefix="../org.oasis-open.xdita.v0_2_2/"/><rewriteURI uriStartString="plugin:org.dita.troff:" rewritePrefix="../org.dita.troff/"/><rewriteURI uriStartString="plugin:org.dita.pdf2.axf:" rewritePrefix="../org.dita.pdf2.axf/"/><rewriteURI uriStartString="plugin:org.dita.html5:" rewritePrefix="../org.dita.html5/"/><rewriteURI uriStartString="plugin:org.lwdita:" rewritePrefix="../org.lwdita/"/><rewriteURI uriStartString="plugin:org.dita.xhtml:" rewritePrefix="../org.dita.xhtml/"/><rewriteURI uriStartString="plugin:org.dita.pdf2:" rewritePrefix="../org.dita.pdf2/"/><rewriteURI uriStartString="plugin:org.oasis-open.dita.v1_2:" rewritePrefix="../org.oasis-open.dita.v1_2/"/><rewriteURI uriStartString="plugin:org.oasis-open.dita.v1_3:" rewritePrefix="../org.oasis-open.dita.v1_3/"/>

       
       <nextCatalog catalog="../org.oasis-open.dita.v1_3/catalog.xml"/><nextCatalog catalog="../org.oasis-open.dita.v1_2/catalog.xml"/><nextCatalog catalog="../org.dita.pdf2/cfg/catalog.xml"/><nextCatalog catalog="../org.dita.specialization.dita11/catalog.xml"/><nextCatalog catalog="../org.oasis-open.xdita.v0_2_2/catalog-dita.xml"/>
      
    </catalog>

[The result of dita-ot-3.3.1/plugins/org.dita.base/catalog-dita.xml with Saxon-HE 9.8.0.15]

    <?xml version="1.0" encoding="UTF-8"?><catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog" xmlns:dita="http://dita-ot.sourceforge.net">

      <rewriteURI uriStartString="platform:" rewritePrefix="../../"/>

      <rewriteURI xmlns="" uriStartString="plugin:org.dita.htmlhelp:" rewritePrefix="../org.dita.htmlhelp/"/><rewriteURI xmlns="" uriStartString="plugin:org.dita.base:" rewritePrefix="../org.dita.base/"/><rewriteURI xmlns="" uriStartString="plugin:org.dita.pdf2.xep:" rewritePrefix="../org.dita.pdf2.xep/"/><rewriteURI xmlns="" uriStartString="plugin:com.sophos.tocjs:" rewritePrefix="../com.sophos.tocjs/"/><rewriteURI xmlns="" uriStartString="plugin:org.dita.pdf2.fop:" rewritePrefix="../org.dita.pdf2.fop/"/><rewriteURI xmlns="" uriStartString="plugin:org.dita.eclipsehelp:" rewritePrefix="../org.dita.eclipsehelp/"/><rewriteURI xmlns="" uriStartString="plugin:org.dita.specialization.dita11:" rewritePrefix="../org.dita.specialization.dita11/"/><rewriteURI xmlns="" uriStartString="plugin:org.dita.normalize:" rewritePrefix="../org.dita.normalize/"/><rewriteURI xmlns="" uriStartString="plugin:org.oasis-open.xdita.v0_2_2:" rewritePrefix="../org.oasis-open.xdita.v0_2_2/"/><rewriteURI xmlns="" uriStartString="plugin:org.dita.troff:" rewritePrefix="../org.dita.troff/"/><rewriteURI xmlns="" uriStartString="plugin:org.dita.pdf2.axf:" rewritePrefix="../org.dita.pdf2.axf/"/><rewriteURI xmlns="" uriStartString="plugin:org.dita.html5:" rewritePrefix="../org.dita.html5/"/><rewriteURI xmlns="" uriStartString="plugin:org.lwdita:" rewritePrefix="../org.lwdita/"/><rewriteURI xmlns="" uriStartString="plugin:org.dita.xhtml:" rewritePrefix="../org.dita.xhtml/"/><rewriteURI xmlns="" uriStartString="plugin:org.dita.pdf2:" rewritePrefix="../org.dita.pdf2/"/><rewriteURI xmlns="" uriStartString="plugin:org.oasis-open.dita.v1_2:" rewritePrefix="../org.oasis-open.dita.v1_2/"/><rewriteURI xmlns="" uriStartString="plugin:org.oasis-open.dita.v1_3:" rewritePrefix="../org.oasis-open.dita.v1_3/"/>

       
       <nextCatalog catalog="../org.oasis-open.dita.v1_3/catalog.xml"/><nextCatalog catalog="../org.oasis-open.dita.v1_2/catalog.xml"/><nextCatalog catalog="../org.dita.pdf2/cfg/catalog.xml"/><nextCatalog catalog="../org.dita.specialization.dita11/catalog.xml"/><nextCatalog catalog="../org.oasis-open.xdita.v0_2_2/catalog-dita.xml"/>
      
    </catalog>

Please indicate me which is the cause of this phenomenon, DITA-OT or Saxon?

My user do needs Saxon-PE with newest DITA-OT, but Saxon-PE 9.8.0.14 is not available in Saxonica Web site. So if Saxon has problem, I want report it to Saxonica.

**Comment by michaelhkay** [comment]
I have reproduced the problem, but it's difficult to debug since it involves understanding a lot of DITA shell script and ant code, neither of which are languages that are easy to debug.

Looking at what changed between 9.8.0.14 and 9.8.0.15, a strong possibility is that we fixed bug #3889 (https://saxonica.plan.io/issues/3889), and DITA is dependent on the incorrect behaviour of Saxon before that bug was fixed.

If that theory is correct, then it will need someone who knows DITA a lot better than I do to identify what DITA change is needed to make this work (but if someone can reduce it to a simple repro then I'll be happy to help). Note that this is logged as Saxon bug 4200 (https://saxonica.plan.io/issues/4200)

Please note that older Saxon releases are generally available for download by constructing a URL that follows the same pattern as published releases, except in the very rare cases where we actively withdraw a release e.g. because of security concerns.

**Comment by drmacro** [comment]
I can take a look at this if nobody else is planning to.

**Comment by michaelhkay** [comment]
If you can identify the transformation that creates the invalid output, and isolate a source document and stylesheet that I can run to reproduce it, then I'll be happy to take it from there.

Michael Kay
Saxonica

**Comment by ToshihikoMakita** [comment]
I have not enough knowledge about DITA-OT processing. However, when I cloned DITA-OT development branch from GitHub, following code seems to put the null namespace `rewriteURI` element into catalog-dita.xml.

[https://github.com/dita-ot/dita-ot/blob/develop/src/main/java/org/dita/dost/platform/ImportPluginInfoAction.java](https://github.com/dita-ot/dita-ot/blob/develop/src/main/java/org/dita/dost/platform/ImportPluginInfoAction.java) Line 51 to 57.

````java
buf.startElement(NULL_NS_URI, "rewriteURI", "rewriteURI", new AttributesBuilder()
    .add("uriStartString", name)
    .add("rewritePrefix", location.toString())
    .build());
buf.endElement(NULL_NS_URI, "rewriteURI", "rewriteURI");
````

**Comment by michaelhkay** [comment]
Ah, the dreaded SAX interface! This has always been a tough one for Saxon because the precise nature of the calls from an XMLReader to a ContentHandler depend on how the XMLReader is configured, and when Saxon doesn't itself create the XMLReader (or other supplier of SAX events) then it has no idea which configuration options to assume. It's also tough because it's a performance-critical interface and Saxon therefore can't afford to do full validation of each call across the interface; if you get it wrong, therefore, the effects may be undefined and can change from release to release.

Here, though, it looks from this fragment as if the application is setting the namespace URI of the rewriteURI element to the "null namespace" so the xmlns="" seems entirely reasonable to me.

There was another bug recently saying that Saxon's behaviour has changed when called via the ContentHandler interface. On that occasion the application was clearly "in breach of contract" as regards Saxon's documentation - it wasn't sending startPrefixMapping() and endPrefixMapping() events. That's a rather weak defence: if you're using the JAXP TransformerHandler interface, Saxon's implementation is supposed to do what the TransformerHandler spec says, rather than making it's own rules, because the application doesn't necessarily even know that it's using Saxon. But unfortunately, the TransformerHandler interface is grossly underspecified. Users are inclined to assume that the JDK implementation of TransformerHandler is somehow definitive and that any difference in behaviour is a bug, but it's hard to be compatible with something that's underspecified.

In practice, the actual effect probably depends on what's on the pipeline behind the ContentHandler, in particular there's often a NamespaceReducer or other component somehere on the pipeline that injects or eliminates namespace declarations, and this is probably where the change occurred between Saxon releases.

Michael Kay
Saxonica

**Comment by ToshihikoMakita** [comment]
For the temporary solution, I have submitted pull-request:

Use OASIS namspace for generating catalog-dita.xml #3287
https://github.com/dita-ot/dita-ot/pull/3287

In my local testing, it works like a charm.
However as this is my first pull-request, it should be verified by DITA-OT contributor.

**Comment by infotexture** [comment]
Fixed for 3.3.2 with #3287.
